images/ci-prune-buildcache/ci_buildcache_prune.py

```diff
@@ -0,0 +1,377 @@
+SINCE_DAYS = os.environ.get("PRUNE_SINCE_DAYS", 14)
+
+# Gitlab API (for fetching artifacts)
+GITLAB_URL = os.environ.get("GITLAB_URL", "https://gitlab.spack.io")
+GITLAB_PROJECT = os.environ.get("GITLAB_PROJECT", "spack/spack")
+gl = gitlab.Gitlab(GITLAB_URL, os.environ.get("GITLAB_TOKEN"))
+
+# S3
+BUILDCACHE_URL = os.environ.get("BUILDCACHE_URL", f"s3://spack-binaries/{PRUNE_REF}/")
+
+logging.basicConfig(level=logging.ERROR)  # Only log ERROR messages
+TODAY = datetime.today()
+
+enable_delete = os.environ.get("AWS_ACCESS_KEY_ID")
+
+
+class SetEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, set):
+            return list(obj)
+        return json.JSONEncoder.default(self, obj)
+
+
+def _project_jobs_gitlabapi(project, ref: str, since=None):
+    if not since:
+        since = datetime.now() - timedelta(days=1)
+
+    for pipeline in project.pipelines.list(iterator=True):
+        created_at = datetime.fromisoformat(pipeline.created_at.replace('Z', '+00:00'))
+
+        # Break if past the since date
+        if created_at.timestamp() < since.timestamp():
+            return
+
+        if not pipeline.ref == ref:
+            continue
+
+        for job in pipeline.jobs.list(iterator=True):
+            if not job.stage == 'generate':
+                continue
+
+            if not job.status == 'success':
+                continue
+
+            yield job
+
+
+# def _project_jobs_postgress(project, ref: str, days: int):
+    # # Gitlab Postgress
+    # db_conn = psycopg2.connect(
+        # host=os.environ.get("GITLAB_PG_HOST"),
+        # port=os.environ.get("GITLAB_PG_PORT"),
+        # dbname=os.environ.get("GITLAB_PG_DBNAME"),
+        # user=os.environ.get("GITLAB_PG_USER"),
+        # password=os.environ.get("GITLAB_PG_PASS"),
+    # )
+    # cur = db_conn.cursor(cursor_factory=RealDictCursor)
+
+    # cur.execute(
+        # f"""
+        # SELECT id
+        # FROM ci_builds
+        # WHERE (
+                # stage = 'generate'
+            # AND status = 'success'
+            # AND project_id = 2
+            # AND (created_at >= CAST((CAST(now() AS timestamp) + (INTERVAL '-{days} day')) AS date))
+            # AND ref = '{ref}'
+        # )
+        # ORDER BY id DESC
+        # """
+    # )
+    # while True:
+        # results = cur.fetchmany()
+        # if not results:
+            # break
+        # for result in results:
+            # yield dict(result)["id"]
+
+
+def fetch_job_hashes(pipeline_job: gitlab.v4.objects.ProjectPipelineJob):
+    jid = pipeline_job.id
+    stack = pipeline_job.name.replace("-generate", "")
+
+    job = project.jobs.get(jid, lazy=True)
+    try:
+        artifact_path = "jobs_scratch_dir/concrete_environment/spack.lock"
+        artifact = job.artifact(artifact_path)
+        lock = json.loads(artifact)
+        return stack, [h for h in lock["concrete_specs"].keys()]
+    except gitlab.exceptions.GitlabHttpError:
+        return []
```

Comment:
Given that you return empty list in this particular exception case, how do you know if/when you're hitting it and that you might not be getting all the data?

Reply:
There are cases when there really is no `spack.lock` file. I originally had this throw an error, but in all of the cases where I could not fetch the lock it simply did not exist.

The assumption here is the Gitlab endpoint is reachable.

Reply:
Ok, I guess that's fine.  But the call site for this method expects to unpack both a stack and list from the return value, so if you end up in here, isn't that going to result in a breaking stop in execution?  I think it will raise `ValueError: too many values to unpack (expected 2)`.

Reply:
ah think you are right! good catch, I will fix that!